### ISSUE TITLE
[Bug] Factorio server startup crash with a lot stored items

### ISSUE THREAD

**Issue by kmteras** [clusterio/clusterio#457]
### Description
While moving items from my main Pyanodon mod server to Clusterio subspace storage, it seems like I have hit a limit on how many items could be stored.

Initially extracting the items was just fine but later when I tried restarting one of the servers, it did not boot up anymore. I narrowed this down to the `subspace_storage` plugin and then my `items.json` state/size. Removing the file and starting with nothing worked. The size of `items.json` was 13328 bytes.

I patched the `items.json` file manually down from 13328 bytes to 10266 bytes and got the servers booting again.

I don't know if this is possible in vanilla Factorio due to the difference in the amount of possible items in vanilla and modded Factorio.

Here is the file that caused the crash if anyone wants to try with Pyanadon mods: [items.txt](https://github.com/clusterio/clusterio/files/8647288/items.txt) (txt due to GitHub's extension limit)

Logs from server startup:
```
  17.612 Verbose Scenario.cpp:390: Map setup finished: 0.832421 seconds.
  17.631 Info UDPSocket.cpp:27: Opening socket at (IP ADDR:({0.0.0.0:34198}))
  17.631 Hosting game at IP ADDR:({0.0.0.0:34198})
  17.631 Info HttpSharedState.cpp:54: Downloading https://auth.factorio.com/generate-server-padlock-2?api_version=4
  17.631 Verbose RouterBase.cpp:60: Started router thread.
  18.050 Info AuthServerConnector.cpp:72: Obtained serverPadlock for serverHash (QDNAUCXGOKyUz0PuGU6425HVjFgNjFKI) from the auth server.
  18.050 Info ServerMultiplayerManager.cpp:794: updateTick(56319845) changing state from(CreatingGame) to(InGame)
  18.050 Info UDPSocket.cpp:39: Opening socket for broadcast
  18.051 Info RemoteCommandProcessor.cpp:130: Starting RCON interface at IP ADDR:({0.0.0.0:65373})
  18.051 Info CommandLineMultiplayer.cpp:292: Maximum segment size = 100; minimum segment size = 25; maximum-segment-size peer count = 10; minimum-segment-size peer count = 20
  18.052 Info RemoteCommandProcessor.cpp:249: New RCON connection from IP ADDR:({127.0.0.1:45534})
[verbose] RCON connection established
[error] Factorio server was unexpectedly shut by down by signal SIGSEGV
[error] Ignoring error from plugin subspace_storage in onStart:
Error: Connection closed
    at EventEmitter.onEnd (/root/clusterio/node_modules/rcon-client/lib/rcon.js:113:45)
    at EventEmitter.emit (events.js:412:35)
    at Socket.<anonymous> (/root/clusterio/node_modules/rcon-client/lib/rcon.js:58:26)
    at Socket.emit (events.js:400:28)
    at TCP.<anonymous> (net.js:686:12)
[error] Ignoring error from plugin server_select in onStart:
Error: Expected state running,stopping but state is init
    at FactorioServer._check (/root/clusterio/node_modules/@clusterio/lib/factorio/server.js:512:10)
    at FactorioServer.sendRcon (/root/clusterio/node_modules/@clusterio/lib/factorio/server.js:902:8)
    at Instance.sendRcon (/root/clusterio/node_modules/@clusterio/slave/slave.js:221:29)
    at InstancePlugin.sendRcon (/root/clusterio/node_modules/@clusterio/lib/plugin.js:233:30)
    at InstancePlugin.sendPendingRcon (/root/clusterio/node_modules/@hornwitser/server_select/instance.js:21:29)
    at InstancePlugin.serialRcon (/root/clusterio/node_modules/@hornwitser/server_select/instance.js:35:9)
    at InstancePlugin.updateList (/root/clusterio/node_modules/@hornwitser/server_select/instance.js:55:14)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async InstancePlugin.onStart (/root/clusterio/node_modules/@hornwitser/server_select/instance.js:65:3
)

**Comment by jake-87** [comment]
Based on the error message, it seems that the factorio server itself was shutdown by something.  Realistically, that item file isn't *that* big ( and isn't on any 2^N bit borders ), so maybe it's an issue with factorio itself? Do you happen to be able to monitor memory usage as it shuts down? The out of memory killer would be my first guess, if you're running on Linux.

**Comment by Danielv123** [comment]
The item loading code should be able to handle up to ~100MB files. Could you post the `factorio-current.log`? It contains all the info as to what actually caused it to crash.

**Comment by kmteras** [comment]
I wanted to make sure that this is not an issue with the server that I was hosting on so I tried and managed to reproduce the errors on my own computer.

For easier reproduction I tested removing all of the mods (besides `clusterio_lib_0.1.2.zip` and `subspace_storage_1.99.8.zip`) but keeping the items list and it still crashes the same way:
```
   0.000 2022-05-10 20:39:02; Factorio 1.1.57 (build 59622, linux64, headless)
   0.022 Operating system: Linux (Ubuntu 22.04)
   0.022 Program arguments: "factorio/1.1.57/bin/x64/factorio" "--config" "instances/test/config.ini" "--start-server" "instances/test/saves/world.zip" "--port" "60225" "--rcon-port" "52448" "--rcon-password" <private> 
   0.022 Read data path: /home/teras/repos/clusterio-test/factorio/1.1.57/data
   0.022 Write data path: /home/teras/repos/clusterio-test/instances/test [386048/1872224MB]
   0.022 Binaries path: /home/teras/repos/clusterio-test/factorio/1.1.57/bin
   0.025 System info: [CPU: AMD Ryzen 7 3800X 8-Core Processor, 16 cores, RAM: 32052 MB]
   0.025 Environment: DISPLAY=:1 WAYLAND_DISPLAY=<unset> DESKTOP_SESSION=ubuntu XDG_SESSION_DESKTOP=ubuntu XDG_CURRENT_DESKTOP=ubuntu:GNOME __GL_FSAA_MODE=<unset> __GL_LOG_MAX_ANISO=<unset> __GL_SYNC_TO_VBLANK=<unset> __GL_SORT_FBCONFIGS=<unset> __GL_YIELD=<unset>
   0.025 Running in headless mode
   0.048 Loading mod core 0.0.0 (data.lua)
   0.073 Loading mod base 1.1.57 (data.lua)
   0.231 Loading mod base 1.1.57 (data-updates.lua)
   0.292 Checksum for core: 1636702648
   0.292 Checksum of base: 1844968960
   0.432 Prototype list checksum: 3034860339
   0.463 Info PlayerData.cpp:71: Local player-data.json available, timestamp 1652204124
   0.463 Info PlayerData.cpp:78: Cloud player-data.json unavailable
   0.464 Info GlobalContext.cpp:1303: Resetting config.
   0.464 Factorio initialised
   0.464 Info ServerSynchronizer.cpp:29: nextHeartbeatSequenceNumber(0) initialized Synchronizer nextTickClosureTick(0).
   0.464 Info ServerMultiplayerManager.cpp:794: updateTick(4294967295) changing state from(Ready) to(PreparedToHostGame)
   0.464 Info ServerMultiplayerManager.cpp:794: updateTick(4294967295) changing state from(PreparedToHostGame) to(CreatingGame)
   0.465 Loading map saves/world.zip: 3412736 bytes.
   0.486 Loading level.dat: 5292557 bytes.
   0.492 Info Scenario.cpp:199: Map version 1.1.57-1
   0.557 Loading script.dat: 9086506 bytes.
   0.561 Checksum for script control.lua: 3902859649
   0.581 Info UDPSocket.cpp:27: Opening socket at (IP ADDR:({0.0.0.0:60225}))
   0.581 Hosting game at IP ADDR:({0.0.0.0:60225})
   0.581 Info HttpSharedState.cpp:54: Downloading https://auth.factorio.com/generate-server-padlock-2?api_version=4
   0.975 Info AuthServerConnector.cpp:72: Obtained serverPadlock for serverHash (Sr2pYw3qT8JcA1u7yaB2I7d1jIcxmxRI) from the auth server.
   0.975 Info ServerMultiplayerManager.cpp:794: updateTick(0) changing state from(CreatingGame) to(InGame)
   0.977 Info ServerRouter.cpp:657: Asking pingpong servers (pingpong1.factorio.com:34197, pingpong2.factorio.com:34197, pingpong3.factorio.com:34197, pingpong4.factorio.com:34197) for own address
   0.977 Info UDPSocket.cpp:39: Opening socket for broadcast
   0.977 Info RemoteCommandProcessor.cpp:130: Starting RCON interface at IP ADDR:({0.0.0.0:52448})
   0.977 Info CommandLineMultiplayer.cpp:292: Maximum segment size = 100; minimum segment size = 25; maximum-segment-size peer count = 10; minimum-segment-size peer count = 20
   0.979 Info RemoteCommandProcessor.cpp:249: New RCON connection from IP ADDR:({127.0.0.1:43462})
[verbose] RCON connection established
2022-05-10 20:39:03 [WARNING] Player <server> tried using the command "rcon.print('disabled')".
2022-05-10 20:39:03 [WARNING] Using Lua console commands will disable achievements. Please repeat the command to proceed.
   1.095 Info ServerRouter.cpp:540: Own address is IP ADDR:({88.196.30.54:60225}) (confirmed by pingpong1)
   1.113 Error ServerMultiplayerManager.cpp:731: Matching server connection failed: Error when creating server game: Missing token.
   1.129 Info ServerRouter.cpp:540: Own address is IP ADDR:({88.196.30.54:60225}) (confirmed by pingpong3)
[error] Factorio server was unexpectedly shut by down by signal SIGSEGV
[error] Ignoring error from plugin subspace_storage in onStart:
Error: Connection closed
    at EventEmitter.onEnd (/home/teras/repos/clusterio-test/node_modules/rcon-client/lib/rcon.js:113:45)
    at EventEmitter.emit (events.js:326:22)
    at Socket.<anonymous> (/home/teras/repos/clusterio-test/node_modules/rcon-client/lib/rcon.js:58:26)
    at Socket.emit (events.js:314:20)
    at TCP.<anonymous> (net.js:675:12)
[error] Ignoring error from plugin server_select in onStart:
Error: Expected state running,stopping but state is init
    at FactorioServer._check (/home/teras/repos/clusterio-test/node_modules/@clusterio/lib/factorio/server.js:512:10)
    at FactorioServer.sendRcon (/home/teras/repos/clusterio-test/node_modules/@clusterio/lib/factorio/server.js:902:8)
    at Instance.sendRcon (/home/teras/repos/clusterio-test/node_modules/@clusterio/slave/slave.js:221:29)
    at InstancePlugin.sendRcon (/home/teras/repos/clusterio-test/node_modules/@clusterio/lib/plugin.js:233:30)
    at InstancePlugin.sendPendingRcon (/home/teras/repos/clusterio-test/node_modules/@hornwitser/server_select/instance.js:21:29)
    at InstancePlugin.serialRcon (/home/teras/repos/clusterio-test/node_modules/@hornwitser/server_select/instance.js:35:9)
    at InstancePlugin.updateList (/home/teras/repos/clusterio-test/node_modules/@hornwitser/server_select/instance.js:55:14)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async InstancePlugin.onStart (/home/teras/repos/clusterio-test/node_modules/@hornwitser/server_select/instance.js:65:3)
    at async Object.timeout (/home/teras/repos/clusterio-test/node_modules/@clusterio/lib/helpers.js:43:10)
```

This also contains all of the logs that `factorio-current.log` has (besides the [verbose] and [error] ones). `1.129 Info ServerRouter.cpp:540:` was the last line logged.

**Comment by kmteras** [comment]
After more testing, it seems like some kind of Factorio RCON limitation or bug? Using the list I had I found the limit where it would and would not crash.

Using [this](https://github.com/clusterio/clusterio/files/8664129/items.txt) as the items database makes even a vanilla server crash for me. But removing the `t` from the last few characters from `"treatedtt":1` to `"treatedt":1` allows the server to boot up fine.

Can anyone confirm that this is reproducible?

**Comment by kmteras** [comment]
I also managed to reproduce this with a more arbitrary input of `{"N":1}` where `N` is any character repeated 13340 times. 13339 does not make it crash.

The file size is different from the other example but that might be due to a difference in encoding.

**Comment by Hornwitser** [comment]
This is due to a bug in Factorio where `/c __anything__ <14k characters>` causes a stack overflow. Not much we can do about it here.

**Comment by kmteras** [comment]
Do you have a link to the bug report for that (if there is one)?

It does not seems like vanilla has enough items for this to occur so I understand if you see it as an edge-case that is not worth working around.

**Comment by Hornwitser** [comment]
There is no forum post about it, but it has been fixed for the next release.

**Comment by Hornwitser** [comment]
Fix has been released in Factorio 1.1.60.
